### ISSUE TITLE
mill-contrib-playlib: java.lang.NoClassDefFoundError: mill/playlib/api/RouteCompilerWorkerApi

### ISSUE THREAD

**Issue by imxieshuang** [com-lihaoyi/mill#907]
```
mill.scalalib.GenIdea.idea java.lang.NoClassDefFoundError: mill/playlib/api/RouteCompilerWorkerApi
    java.lang.ClassLoader.defineClass1(Native Method)
    java.lang.ClassLoader.defineClass(ClassLoader.java:756)
    java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
    java.net.URLClassLoader.defineClass(URLClassLoader.java:468)
    java.net.URLClassLoader.access$100(URLClassLoader.java:74)
    java.net.URLClassLoader$1.run(URLClassLoader.java:369)
    java.net.URLClassLoader$1.run(URLClassLoader.java:363)
    java.security.AccessController.doPrivileged(Native Method)
    java.net.URLClassLoader.findClass(URLClassLoader.java:362)
    mill.api.ClassLoader$$anon$1.findClass(ClassLoader.scala:23)
    java.lang.ClassLoader.loadClass(ClassLoader.java:418)
    java.lang.ClassLoader.loadClass(ClassLoader.java:351)
    mill.playlib.RouteCompilerWorker.bridge(RouteCompilerWorkerApi.scala:27)
    mill.playlib.RouteCompilerWorker.compile(RouteCompilerWorkerApi.scala:48)
    mill.playlib.RouterModule.$anonfun$compileRouter$2(RouterModule.scala:62)
    mill.define.ApplyerGenerated.$anonfun$zipMap$3(ApplicativeGenerated.scala:9)
    mill.define.Task$MappedDest.evaluate(Task.scala:392)
    mill.eval.Evaluator.$anonfun$evaluateGroup$13(Evaluator.scala:476)
    scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
    scala.Console$.withErr(Console.scala:193)
    mill.eval.Evaluator.$anonfun$evaluateGroup$12(Evaluator.scala:476)
    scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
    scala.Console$.withOut(Console.scala:164)
    mill.eval.Evaluator.$anonfun$evaluateGroup$11(Evaluator.scala:475)
    scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
    scala.Console$.withIn(Console.scala:227)
    mill.eval.Evaluator.$anonfun$evaluateGroup$8(Evaluator.scala:474)
    mill.eval.Evaluator.$anonfun$evaluateGroup$8$adapted(Evaluator.scala:426)
    scala.collection.immutable.Vector.foreach(Vector.scala:1794)
    mill.eval.Evaluator.evaluateGroup(Evaluator.scala:426)
    mill.eval.Evaluator.evaluateGroupCached(Evaluator.scala:329)
    mill.eval.Evaluator.$anonfun$sequentialEvaluate$2(Evaluator.scala:110)
    scala.collection.IterableOnceOps.foreach(IterableOnce.scala:553)
    scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:551)
    scala.collection.AbstractIterator.foreach(Iterator.scala:1279)
    mill.eval.Evaluator.sequentialEvaluate(Evaluator.scala:93)
    mill.eval.Evaluator.evaluate(Evaluator.scala:79)
    mill.scalalib.GenIdeaImpl.$anonfun$xmlFileLayout$90(GenIdeaImpl.scala:379)
    scala.collection.immutable.Vector1.map(Vector.scala:1824)
    scala.collection.immutable.Vector1.map(Vector.scala:348)
    mill.scalalib.GenIdeaImpl.xmlFileLayout(GenIdeaImpl.scala:373)
    mill.scalalib.GenIdeaImpl.run(GenIdeaImpl.scala:36)
    mill.scalalib.GenIdea$.$anonfun$idea$2(GenIdea.scala:14)
    scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
```

**Comment by lefou** [comment]
Can you post your `build.sc` or some minimized repro?

**Comment by imxieshuang** [comment]
import coursier.maven.MavenRepository
import mill._
import scalalib._
import scalajslib._
import publish._
import mill.modules.Jvm.createAssembly
import $ivy.`com.lihaoyi::mill-contrib-playlib:$MILL_VERSION`
import mill.playlib._

object server extends PlayModule {

  //config
  override def scalaVersion = T {
    "2.12.10"
  }

  override def playVersion = T {
    "2.7.5"
  }

  override def twirlVersion = T {
    "1.4.2"
  }

  override def ivyDeps = T {
    super.ivyDeps() ++ Agg(ws(), filters())
  }

  object test extends PlayTests

}

**Comment by imxieshuang** [comment]
This is mill 0.7.3, it is ok at 0.6x

**Comment by lefou** [comment]
Does your project compile successfully? `mill __.compile`

**Comment by lefou** [comment]
I can't reproduce your issue. I can successfully generate a IDEA project with mill 0.7.3 and the above posted `build.sc`.

**Comment by svefors** [comment]
I've seen it too, or at least I think it is the same classloader issue.
Using 0.7.3
On Darwin
Using Zulu Java 8, 11, 14
./mill core.compileRouter

Exception in thread "MillServerActionRunner" java.lang.NoClassDefFoundError: mill/playlib/api/RouteCompilerWorkerApi
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1017)
	at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:151)
	at java.base/java.net.URLClassLoader.defineClass(URLClassLoader.java:514)
	at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:422)
	at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:416)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:691)
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:415)
	at mill.api.ClassLoader$$anon$1.findClass(ClassLoader.scala:23)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:589)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	at mill.playlib.RouteCompilerWorker.bridge(RouteCompilerWorkerApi.scala:27)
	at mill.playlib.RouteCompilerWorker.compile(RouteCompilerWorkerApi.scala:48)
	at mill.playlib.RouterModule.$anonfun$compileRouter$2(RouterModule.scala:62)
	at mill.define.ApplyerGenerated.$anonfun$zipMap$3(ApplicativeGenerated.scala:9)
	at mill.define.Task$MappedDest.evaluate(Task.scala:392)
	at mill.eval.Evaluator.$anonfun$evaluateGroup$13(Evaluator.scala:476)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
	at scala.Console$.withErr(Console.scala:193)
	at mill.eval.Evaluator.$anonfun$evaluateGroup$12(Evaluator.scala:476)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
	at scala.Console$.withOut(Console.scala:164)
	at mill.eval.Evaluator.$anonfun$evaluateGroup$11(Evaluator.scala:475)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
	at scala.Console$.withIn(Console.scala:227)
	at mill.eval.Evaluator.$anonfun$evaluateGroup$8(Evaluator.scala:474)
	at mill.eval.Evaluator.$anonfun$evaluateGroup$8$adapted(Evaluator.scala:426)
	at scala.collection.immutable.Vector.foreach(Vector.scala:1794)
	at mill.eval.Evaluator.evaluateGroup(Evaluator.scala:426)
	at mill.eval.Evaluator.evaluateGroupCached(Evaluator.scala:329)
	at mill.eval.Evaluator.$anonfun$sequentialEvaluate$2(Evaluator.scala:110)
	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:553)
	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:551)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1279)
	at mill.eval.Evaluator.sequentialEvaluate(Evaluator.scala:93)
	at mill.eval.Evaluator.evaluate(Evaluator.scala:79)
	at mill.main.RunScript$.evaluate(RunScript.scala:202)
	at mill.main.RunScript$.$anonfun$evaluateTasks$1(RunScript.scala:188)
	at scala.util.Either.map(Either.scala:382)
	at mill.main.RunScript$.evaluateTasks(RunScript.scala:187)
	at mill.main.RunScript$.$anonfun$runScript$4(RunScript.scala:66)
	at ammonite.util.Res$Success.flatMap(Res.scala:62)
	at mill.main.RunScript$.runScript(RunScript.scala:65)
	at mill.main.MainRunner.$anonfun$runScript$1(MainRunner.scala:115)
	at mill.main.MainRunner.watchLoop2(MainRunner.scala:64)
	at mill.main.MainRunner.runScript(MainRunner.scala:90)
	at mill.MillMain$.main0(MillMain.scala:239)
	at mill.main.MillServerMain$.main0(MillServerMain.scala:67)
	at mill.main.Server.$anonfun$handleRun$1(MillServerMain.scala:154)
	at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: java.lang.ClassNotFoundException: mill.playlib.api.RouteCompilerWorkerApi
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:602)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	at mill.api.ClassLoader$$anon$1.findClass(ClassLoader.scala:22)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:589)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)

**Comment by lefou** [comment]
@svefors Your not even trying to generate the IDEA files. It looks to me the issue is in the playlib contrib module. If I run the same command with mill 0.7.3 on the above `build.sc`, I get a different error. I guess I still need more info to reproduce the issue:

```
10:27:49 θ60° lefou:/tmp/mill-playlib 3s $ mill -i __.compileRouter
Compiling /tmp/mill-playlib/build.sc
[10/11] server.routeFiles 
1 targets failed
server.routeFiles java.nio.file.NoSuchFileException: /tmp/mill-playlib/server/conf
    os.walk$stream$.attrs(ListOps.scala:198)
    os.walk$stream$.apply(ListOps.scala:161)
    os.walk$.apply(ListOps.scala:95)
    mill.playlib.RouterModule.$anonfun$routeFiles$3(RouterModule.scala:16)
    scala.collection.StrictOptimizedIterableOps.flatMap(StrictOptimizedIterableOps.scala:117)
    scala.collection.StrictOptimizedIterableOps.flatMap$(StrictOptimizedIterableOps.scala:104)
    scala.collection.immutable.Vector.flatMap(Vector.scala:113)
    mill.playlib.RouterModule.$anonfun$routeFiles$2(RouterModule.scala:16)
    mill.define.Task$MappedDest.evaluate(Task.scala:392)
```

**Comment by lefou** [comment]
But of course, generating the IDEA files works.

```
10:30:53 lefou:/tmp/mill-playlib 130 $ mill -i mill.scalalib.GenIdea/idea
[1/1] mill.scalalib.GenIdea.idea 
Analyzing modules ...
[36/37] server.test.ideaCompileOutput 
Writing IDEA project files ...
```

**Comment by svefors** [comment]
> @svefors This seems to be unrelated to generating the IDEA files. It looks to me the issue is in the playlib contrib module. If I run the same command with mill 0.7.3 on the above `build.sc`, I get a different error though. I guess I still need more info to reproduce the issue:
> 
> ```
> 10:27:49 θ60° lefou:/tmp/mill-playlib 3s $ mill -i __.compileRouter
> Compiling /tmp/mill-playlib/build.sc
> [10/11] server.routeFiles 
> 1 targets failed
> server.routeFiles java.nio.file.NoSuchFileException: /tmp/mill-playlib/server/conf

I think you have to add the play directories to the project.

**Comment by svefors** [comment]
> @svefors This seems to be unrelated to generating the IDEA files. It looks to me the issue is in the playlib contrib module. 

I think you are right about the playlib contrib module. I see some change in there related to an upgrade of Mill to scala 2.13. There is a comment in the pull request that since Play 2.6 and 2.7 don't support scala 2.13 it will work in a different way than the other modules.

"Only scalanativelib and playlib cannot be upgraded to 2.13 due to dependencies; we leave those in 2.12, and turn their respective *.api modules into pure Java modules so they can be shared between both the Scala 2.12 *Worker an the Scala 2.13 *Module modules"

the "mill/playlib/api/RouteCompilerWorkerApi" was a scala file but is now a java.

**Comment by lefou** [comment]
I can confirm that mill 0.7.3 fails for the given project.

```
08:48:09 lefou:/tmp/mill-play-sample master ± mill --mill-version 0.6.2 __.compile
[77/77] core.test.compile 
08:48:13 lefou:/tmp/mill-play-sample master 2s ± mill --mill-version 0.6.2 clean
[1/1] clean 
08:48:18 lefou:/tmp/mill-play-sample master ± mill --mill-version 0.7.3 __.compile
Mill version changed (0.6.2 -> 0.7.3), re-starting server
Compiling /tmp/mill-play-sample/build.sc
[18/76] core.compileRouter 
Exception in thread "MillServerActionRunner" java.lang.NoClassDefFoundError: mill/playlib/api/RouteCompilerWorkerApi
        at java.lang.ClassLoader.defineClass1(Native Method)
        at java.lang.ClassLoader.defineClass(ClassLoader.java:756)
        at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
        at java.net.URLClassLoader.defineClass(URLClassLoader.java:468)
        at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
        at java.net.URLClassLoader$1.run(URLClassLoader.java:369)
        at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
        at mill.api.ClassLoader$$anon$1.findClass(ClassLoader.scala:23)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
        at mill.playlib.RouteCompilerWorker.bridge(RouteCompilerWorkerApi.scala:27)
        at mill.playlib.RouteCompilerWorker.compile(RouteCompilerWorkerApi.scala:48)
        at mill.playlib.RouterModule.$anonfun$compileRouter$2(RouterModule.scala:62)
        at mill.define.ApplyerGenerated.$anonfun$zipMap$3(ApplicativeGenerated.scala:9)
        at mill.define.Task$MappedDest.evaluate(Task.scala:392)
        at mill.eval.Evaluator.$anonfun$evaluateGroup$13(Evaluator.scala:476)
        at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
        at scala.Console$.withErr(Console.scala:193)
        at mill.eval.Evaluator.$anonfun$evaluateGroup$12(Evaluator.scala:476)
        at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
        at scala.Console$.withOut(Console.scala:164)
        at mill.eval.Evaluator.$anonfun$evaluateGroup$11(Evaluator.scala:475)
        at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
        at scala.Console$.withIn(Console.scala:227)
        at mill.eval.Evaluator.$anonfun$evaluateGroup$8(Evaluator.scala:474)
        at mill.eval.Evaluator.$anonfun$evaluateGroup$8$adapted(Evaluator.scala:426)
        at scala.collection.immutable.Vector.foreach(Vector.scala:1794)
        at mill.eval.Evaluator.evaluateGroup(Evaluator.scala:426)
        at mill.eval.Evaluator.evaluateGroupCached(Evaluator.scala:329)
        at mill.eval.Evaluator.$anonfun$sequentialEvaluate$2(Evaluator.scala:110)
        at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:553)
        at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:551)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1279)
        at mill.eval.Evaluator.sequentialEvaluate(Evaluator.scala:93)
        at mill.eval.Evaluator.evaluate(Evaluator.scala:79)
        at mill.main.RunScript$.evaluate(RunScript.scala:202)
        at mill.main.RunScript$.$anonfun$evaluateTasks$1(RunScript.scala:188)
        at scala.util.Either.map(Either.scala:382)
        at mill.main.RunScript$.evaluateTasks(RunScript.scala:187)
        at mill.main.RunScript$.$anonfun$runScript$4(RunScript.scala:66)
        at ammonite.util.Res$Success.flatMap(Res.scala:62)
        at mill.main.RunScript$.runScript(RunScript.scala:65)
        at mill.main.MainRunner.$anonfun$runScript$1(MainRunner.scala:115)
        at mill.main.MainRunner.watchLoop2(MainRunner.scala:64)
        at mill.main.MainRunner.runScript(MainRunner.scala:90)
        at mill.MillMain$.main0(MillMain.scala:239)
        at mill.main.MillServerMain$.main0(MillServerMain.scala:67)
        at mill.main.Server.$anonfun$handleRun$1(MillServerMain.scala:154)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassNotFoundException: mill.playlib.api.RouteCompilerWorkerApi
        at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
        at mill.api.ClassLoader$$anon$1.findClass(ClassLoader.scala:22)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
        ... 51 more
```

**Comment by dkovalenko** [comment]
Any updates? Play framework cannot be built using Mill now because of this issue 😒

**Comment by lefou** [comment]
There where PRs trying to fix this in the past. Unfortunately all went silent, but some were reviewed with useful info how to fix it.

**Comment by lefou** [comment]
Fixed by #1324

**Comment by lefou** [comment]
Issue still not properly fixed, but I create a new PR #1330.
